accepted/2024/trusted-publishers-oidc-for-nuget-push.technical.md

```diff
@@ -0,0 +1,253 @@
+  - `environment` claim must be `{environment}` (case insensitive)
+- If a workflow path filter is provided in the trust policy:
+   - `job_workflow_ref` claim must be `{repo owner}/{repo name}/{workflow path}@.*` (case insensitive)
+   - The workflow path should be normalized to `/` path separators at the time of trust policy creation, by the package
+     source, for better UX
+- If a tag filter is provided in the trust policy:
+  - `ref_type` claim must be `tag`
+  - `ref` claim must match `refs/tags/{tag-pattern}` (case sensitive, tag names of differing case can coexist, tag match supports wildcards)
+
+A list of possible claims to verify against is available in GitHub's [Understanding the OIDC
+token](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#understanding-the-oidc-token)
+document.
+
+Other Trusted Publishers like Azure DevOps or Bitbucket should have sufficient token claims so both the registry
+(NuGet.org) and the package author are certain that only proper workload identity tokens are traded for privileged
+short-lived API keys.
+
+## Data relationships (persisting trust policies, schema changes)
+
+A new SQL table will be added to the NuGetGallery database to store trust policies. The table should be generic enough
+to allow us to add additional Trusted Publishers without a DB schema change (ideally). The trust policy DB record with
+have a foreign key to the `Users` table (containing both user and organization records) but will be restricted from
+associating with organization records by the application. Many of the columns will be shared with the `Credentials`
+table in order to express the scoping rules to be copied into the short-lived API key.
+
+A nullable column will be added to the `Credentials` table to refer to the trust policy used to create the short-lived
+API key. This will allow package publish operations to audit their related trust policy information.
+
+Deleting a trust policy should have the effect of deleting all related short-lived API keys.
+
+## Trading an OIDC token for an API key
+
+A new endpoint will be needed for trading an bearer token (OIDC token, a JWT) for an API key. The endpoint URL will be
+discoverable via the [V3 service index](https://learn.microsoft.com/en-us/nuget/api/overview#service-index) and a new
+resource type which is `TokenService/1.0.0`. For NuGet.org, the service index is available at
+`https://api.nuget.org/v3/index.json`. The new resource URL be something like `https://www.nuget.org/api/v2/token`.
+
+This new resource will be referred to as the **token service**. This generic name will allow for future dynamic
+authorization scenarios beyond trading a bearer token for a NuGet API key.
+
+```
+POST /api/v2/token HTTP/1.1
+Host: www.nuget.org
+Authorization: Bearer {OIDC token}
+Content-Type: application/json
+
+{
+   "username": "{username of user with trust policy}"
+}
```

Comment:
Mostly curious: why is `username` necessary here? Shouldn't the OIDC token always map 1:1 with a configuration/policy?

Reply:
In this design a trust policy is attached to a user. Inside the trust policy, the user specifies which user/org the trust policy is scoped to (to handle the `“Pending” Trusted Publishers` problem mentioned in Seth's doc) -- a user can publish packages as themselves or to an organization they are a memeber of. So this `username` parameter specifies the user profile than holds the trust policy. 

There is no enforcement of 1:1 mapping planned. If we did want to enforce a 1:1 mapping we'd have to decide who (which NuGet.org user profile) gets preference. A first come first service basis would be weird too -- who would win?

I think is it better to allow users to set up whatever trust policy they want and require this additional parameter during token trade. The token trade would first fetch all trust policies for that username and see if any of them apply to the incoming token.

As an added benefit this would allow a future model where the trust policy is attached to an organization directly (instead of one of the admins). But this is not planned yet and we're more approaching this as a way to mint short-lived API keys which are only owned by users today.

Reply:
Got it, so the configuration of the `nuget/login` would need to include a single username as well, correct? I'm imagining that it might be hard to determine who this user would be in the case of a project with multiple maintainers. Does this also mean that the projects other maintainers have no way to administer the publisher?

(For context, in PyPI's implementation, publishers map to projects, not users, and all owners/maintainers of a given project can change/configure publishers for that project. But that does require 'pending publishers', which is a little awkward UX-wise, but seems to work)

Reply:
Thank you for testing my logic here! I really appreciate it!

Yes, `nuget/login` would have a username provided. I have a bit of sample YAML in the spec (exact shape TBD):
```yaml
      - name: NuGet login
        uses: nuget/login@v1
        id: login
        with:
          user: joelverhagen # this is the user profile which has the GitHub Actions trust policy
          source: https://api.nuget.org/v3/index.json
```

> I'm imagining that it might be hard to determine who this user would be in the case of a project with multiple maintainers.

Yes, this is not great. But it is the current problem we have with "who owns the long lived" API key, so it's not worse than what we have today. If a user is concerned about sharing their user profile name it could be a GitHub secret variable `NUGET_USERNAME`. 

> Does this also mean that the projects other maintainers have no way to administer the publisher?

The term `publisher` is not well defined in NuGet world, but when used that typically refers to the visible owner (user/org) of the package. But I think based on that PyPI definition I think that a PyPI project is equivalent to a NuGet.org package ID (sometimes called package registration). I don't think we would want to attach trust policies to a package ID since orgs have many many package IDs (hundreds) and they would need to set the policy on all those places (painful). I guess .NET people like a bunch of little packages for some reason 😆.

Suppose a core contributor on a GitHub repo disappears and another contributor needs to publish a new version of the package(s). They would add an equivalent trust policy to their user account scoped to the org (or user co-owner) that owns the GitHub repo's packages and then specify their username in the token trade as we discussed above.

I think this is the PyPI + NuGet term mapping (did this mainly to get my head straight)

PyPI | NuGet.org | Notes 
--- | --- | ---
Project | Package ID / Package Registration
Release | Package version | Sometimes just called "package" in NuGet, confusingly
Collaborator / Maintainer (?) | Owner | Can be a user or organization on NuGet.org, not sure about PyPI
Organization | Organization | Looks like creating these has an approval flow on PyPI, unlike NuGet.org
User | User | These seem to have equivalent usages